### ISSUE TITLE
Fix Error 1002 missing schema on project installer.

### ISSUE THREAD

**Issue by yorkshire-pudding** [backdrop/backdrop-issues#5690]
## Description of the bug
Enabled project installer on a site that had been upgraded a while ago from D7 but have always used `bee` to download and install.
I don't think project installer was enabled on the D7 site so an update might have been missed and may not have the projects server listed..  I don't get this anywhere else.
Got error (that looks like success message) and no projects listed

This is may not be a Backdrop bug and may be related to the upgrade process and the unusual circumstances of not having project installer enabled, but would appreciate guidance on how to fix it.  Happy to apply SQL but don't know where to look.

## Steps To Reproduce
Not sure - perhaps start with D7 site with project installer disabled
Then upgrade
Then enable project installer 

## Actual behavior

No results returned.  Notices, warnings and error message.  

![image](https://user-images.githubusercontent.com/20357797/179976038-953cfd54-84f3-4e28-b671-71d275ae16f2.png)

<details>
Warning: array_merge(): Expected parameter 2 to be an array, null given in installer_browser_fetch_results() (line 240 of /app/docroot/core/modules/installer/installer.browser.inc).
Notice: Trying to access array offset on value of type null in installer_browser_fetch_results() (line 248 of /app/docroot/core/modules/installer/installer.browser.inc).
Notice: Trying to access array offset on value of type null in installer_browser_fetch_results() (line 256 of /app/docroot/core/modules/installer/installer.browser.inc).
[Dismiss](https://iji-bd.lndo.site/admin/modules/install#)
Status messageEncountered an error when trying to fetch results from . Error -1002 : missing schema
</details>

Also can't install by manual > install projects by name

![image](https://user-images.githubusercontent.com/20357797/179977533-3dcd453e-856d-4873-85be-6f4168a635be.png)

<details>
Warning: array_merge(): Expected parameter 2 to be an array, null given in installer_browser_fetch_results() (line 240 of /app/docroot/core/modules/installer/installer.browser.inc).
Notice: Trying to access array offset on value of type null in installer_browser_fetch_results() (line 248 of /app/docroot/core/modules/installer/installer.browser.inc).
Notice: Trying to access array offset on value of type null in installer_browser_fetch_results() (line 256 of /app/docroot/core/modules/installer/installer.browser.inc).
Notice: Undefined index: projects in installer_manager_install_form_validate() (line 688 of /app/docroot/core/modules/installer/installer.manager.inc).
Sorry, the project server is offline.
Notice: Trying to access array offset on value of type null in installer_manager_install_form_validate() (line 723 of /app/docroot/core/modules/installer/installer.manager.inc).
Notice: Trying to access array offset on value of type null in installer_manager_install_form_validate() (line 726 of /app/docroot/core/modules/installer/installer.manager.inc).
Notice: Trying to access array offset on value of type null in installer_manager_install_form_validate() (line 744 of /app/docroot/core/modules/installer/installer.manager.inc).
[Dismiss](https://iji-bd.lndo.site/admin/installer/manual#)
Status messageEncountered an error when trying to fetch results from . Error -1002 : missing schema
</details>

## Expected behavior

Project installer lists projects and allows install by name

## Additional information

Add any other information that could help, such as:
- Backdrop CMS version: 1.22
- Web server and its version: apache
- Operating System and its version: n/a
- Browser(s) and their versions: n/a

**Comment by indigoxela** [comment]
Very odd. Can you check the following:

- Does the problem persist when you flush all caches?
- Does the config file `installer.settings.json` exist in your active config directory?
- If it does, what are the contents?

Background to this: the first problem seems that in `array_merge($filters, $server)` the $server variable seems to be null, but there's a default setting, so it at least contains `https://projects.backdropcms.org`.

**Comment by argiepiano** [comment]
The problem is not likely related to the upgrade. In D7, the installer is a module called `update`, which doesn't exist in Backdrop, where the installer is called `installer`. In fact, no "D7 to Backdrop" upgrade ever has the `installer` module on after the upgrade (it's one of those modules that don't get automatically enabled) - you always have to enable it manually after the upgrade, regardless of whether the D7 had `update` on or off. So you can most likely discard the upgrade process (in fact `installer.install` has only one hook_update_N(), which is not related to the upgrade process). 

As @indigoxela pointed out, the most likely culprit is a problem with `installer.settings.json`. Inspect that file both in your site's `files` folder and in the `core/modules/installer/config` location.

Also, you may have a files directory problem. Check the site's `settings.php` file (and any possible `settings.local.php` file) for the line `$config_directories['active']` and see where it points, then be sure that   `installer.settings.json` is present in that folder.

**Comment by argiepiano** [comment]
> This update hook config change should IMO not run, if the installer module isn't on.

I agree. In fact, I've upgraded dozens of sites while testing other PRs and this hook has never before run. I think it's safe to remove this hook, as there is no "update" from a D7 version, as `installer` doesn't exist in D7 (update_1000() is used for upgrades from D7). 

I've done more testing to verify: I can confirm that enabling Installer in a freshly upgraded D7->B site does not run `installer_update_1000()`. I've tested this via en UI and via bee. I haven't  tested with drush.

Bottom line: I can't explain if or why this hook ran in @yorkshire-pudding site. But it's probably a good idea to leave this open in case others run into this issue.

**Comment by indigoxela** [comment]
> I think it's safe to remove this hook, as there is no "update" from a D7 version, as installer doesn't exist in D7 (update_1000() is used for upgrades from D7).

Looking at git blame - this update hook came in with #3105 and still seems OK. Its main intention was to prevent enabling core updates for B sites installed before the core update was enabled.

There are two situations, though, where this hook does harm:

1. Upgrading from D7
2. Very, very old B sites (older than 1.11.0) that have the installer disabled

So, why does this stub config problem exist.
You can verify with following code:

```
$config = config('foobar');
debug($config);
$config->set('bla', 'blu');
debug($config);
$config->save();// Problem!
```

The fix can be a one-liner, we only have to check if `$config-isNew()` and in that case do nothing.

**Comment by indigoxela** [comment]
A [PR](https://github.com/backdrop/backdrop/pull/4141) is available for testing and review.

**How to test:**

Update from D7 and verify there's no stub installer.settings.json file in the active config dir
or
Update from a B install older than 1.11.0 - in that case: if the installer was on, the core updates should be off in config. If the installer was off, no stub file gets created in the active config dir.

I hope, my explanation is clear enough (I often fail to properly explain things) :wink:.

**Comment by yorkshire-pudding** [comment]
Thanks @indigoxela.  I confirm:
- after running `core/update.php` there is no `installer.settings.json`
- After enabling Project Installer it is there and module works as expected.

**Comment by argiepiano** [comment]
Aha! I see that update hook was added as per [this comment](https://github.com/backdrop/backdrop-issues/issues/3105#issuecomment-417819553) in that issue. 3105  Thanks @indigoxela. (I did my homework and read the whole thread this time 😉 ) 

Your fix makes sense. But your explanations for testing make only partial sense to me:

> Upgrade from D7 and verify there's no stub installer.settings.json file in the active config dir

This doesn't make sense to me. Before your patch, none of the D7 upgrades I'm doing ever run `installer_update_1000()`, and none of them ever create a "stub" `installer.settings.json` file - neither before, nor after your patch. This is because `installer` doesn't exist in D7, and therefore it is always disabled when upgrading from D7. And when you manually enable `installer` with the current version of Backdrop after the upgrade, the update hook is not run (which is the expected behavior).

As I said above, the D7 upgrade **itself** was probably never the problem. It alone can't explain @yorkshire-pudding  problem. There must have been an additional problem in between. 

> or
Update from a B install older than 1.11.0 - in that case: if the installer was on, the core updates should be off in config. If the installer was off, no stub file gets created in the active config dir.

So, this is also not clear to me.

I did try this:
- Started with version 1.10.2. (Backdrop looked a bit different back then! 😄 )
- I  manually turned off the installer. 
- I upgrade to 1.11.0. No "stub" file is created. Simply: the update hook is not run because `installer` is off, and the stub json file is not created.

Then I tried this:
- Started with version 1.10.2.
- I made sure installer is ON. The file `installer.settings.json` file exists normally "sans" the `core_update` setting
- Updated to 1.11.0. The `installer.settings.json` file is updated (the update DOES run this time) and `"core_update" : false` is added normally. The rest of the json file is left untouched - it's not a stub file, but a beautifully looking full settings file.

So none of these situations can explain @yorkshire-pudding problem either.

My theory is that something went wrong with @yorkshire-pudding's site, and that it was really a fringe case. Therefore fixing `installer.install` with your patch may not be necessary.  Your patch doesn't do any harm and is very simple, but it may not be needed in normal situations.  Still... better safe than sorry, so I'm going to marked this as RTBC.

So this is my theory of what happened to @yorkshire-pudding:
- He probably updated from D7 before 1.11.0. But this fact is probably irrelevant as I explained above.
- He enabled `installer` in his pre 1.11.0 version in a way that did not move `installer.settings.json` to the config folder (perhaps manually switching it on in the database?), OR that json file was somehow removed from the config directory. So he ended up with an enabled `installer` and no json file, **somehow**. 
- He then updated Backdrop to 1.11.0, and this ran `installer_update_1000()`. But since there was no json file in the config folder, then the empty stub was created

**Comment by indigoxela** [comment]
@argiepiano thank you so much for your thorough testing! :pray: 

You're absolutely right, the problem must be something else - and I found a way to reproduce it on a fresh vanilla install:

1. Disable and uninstall the installer module (json file gone)
2. Go to /admin/reports/updates/settings
3. Save the file - at this point the stub json file gets created
4. Enable the installer module again - the file stays broken

So the bug is indeed somewhere else: it's in function update_settings() - the form still displays the core update setting, even if the installer module isn't enabled.

**Comment by indigoxela** [comment]
Never too late to switch strategy. :wink:

A totally different [PR](https://github.com/backdrop/backdrop/pull/4142) is available for testing and review. Sorry to waste your (testing) time.

This time testing is much easier (can also happen in the sandbox). Just follow the steps in my previous comment.

**Comment by argiepiano** [comment]
> So the bug is indeed somewhere else: it's in function update_settings()

Wow! That explains the problem, and it's indeed something that can happen much more often. I think the problem is that the option to enable the Manual self update should not be shown if installer is disabled, right?

I see you just submitted a new PR. I'll take a look

**Comment by yorkshire-pudding** [comment]
> @argiepiano thank you so much for your thorough testing! 🙏
> 
> You're absolutely right, the problem must be something else - and I found a way to reproduce it on a fresh vanilla install:
> 
> 1. Disable and uninstall the installer module (json file gone)
> 2. Go to /admin/reports/updates/settings
> 3. Save the file - at this point the stub json file gets created
> 4. Enable the installer module again - the file stays broken
> 
> So the bug is indeed somewhere else: it's in function update_settings() - the form still displays the core update setting, even if the installer module isn't enabled.

That makes sense and is almost certainly what happened.
Thanks - will do test

**Comment by argiepiano** [comment]
This new PR looks great and WFM. All tests passing.   RTBC. Perhaps we should change the title of this issue to reflect the new direction? 

Thanks for digging into this and finding the culprit, @indigoxela!

**Comment by argiepiano** [comment]
One more thought. 

For sites that currently present the problem described by @yorkshire-pudding in this issue (because Project Installer was disabled, and they visited and saved the form at  `/admin/reports/updates/settings`):  this patch with not fix the already problem created by the faulty code. The stub json file already exists in the config directory.  

In fact there are only two ways to fix the problem once it's created: (1) manually fix the json file as @yorkshire-pudding did.  (2) Or you would need to enable and then disable `installer` to fully remove the faulty json file.

Should we provide another upgrade hook that checks that the json file exists, and if it contains the key `"installer_server"`,  and if not, creates the default  installer_server entry?

**Comment by yorkshire-pudding** [comment]
I've tested the PR and works.  Thanks @indigoxela you are a star.
@argiepiano - good point, though not sure how common it will be.

**Comment by yorkshire-pudding** [comment]
@argiepiano @indigoxela 
Would a suitable title be: "Project installer doesn't work if settings are changed while disabled"
?

**Comment by argiepiano** [comment]
> good point, though not sure how common it will be.

It can be common. If you upgrade a D7 site, Installer is disabled by default. If you visit `/admin/reports/updates/settings` and save that form, the faulty json file is created. At this point the only way to fix the problem is to manually alter the json file as @yorkshire-pudding did, OR installing - fully uninstalling - installing again Project Installer so that the uninstall step removes the faulty file.

This update hook is simple. It would check if a faulty json config file exists, and restore the default value. It could look like this

```php
function installer_update_1001() {
  $config = config('installer.settings');
  if (!$config->isNew() && empty($config->get('installer_server'))) {
    $config->set('installer_server', array(
      'url' => 'https://projects.backdropcms.org',
      'name' => 'Backdrop',
    ));
    $config->save();
  }
}
```

**Comment by quicksketch** [comment]
Thanks @indigoxela, @yorkshire-pudding, and @argiepiano! I merged https://github.com/backdrop/backdrop/pull/4142 since it stands alone as a great fix.

@argiepiano I think your idea to make a "fix bad JSON" update might also help be comprehensive. Do you think we should add that as well?

**Comment by argiepiano** [comment]
> @argiepiano I think your idea to make a "fix bad JSON" update might also help be comprehensive. Do you think we should add that as well?

Perhaps on a separate issue, for 1.23.0, since this was merged already? Or is there time to do this for 1.22.2? I can probably open a new issue and PR in the next 3 or 4 hours

**Comment by quicksketch** [comment]
Yeah let's do it as a new issue. Doesn't necessarily need to wait for
1.23.0 though. We could get it in 1.22.3.

**Comment by indigoxela** [comment]
Follow-up created: #5696
